server/src/main/java/io/crate/expression/scalar/ArraySumFunction.java

```diff
@@ -0,0 +1,151 @@
+            }
+        }
+        return sum;
+    };
+
+    private final Function<List<Double>, Double> doubleFunction = doubles -> {
+        var kahanSummationForDouble = new KahanSummationForDouble();
+        double sum = 0;
+        for (int i = 0; i < doubles.size(); i++) {
+            var item = doubles.get(i);
+            if (item != null) {
+                sum = kahanSummationForDouble.sum(sum, item);
+            }
+        }
+        return sum;
+    };
+
+    private final Function<List<Number>, Long> longFunction = list -> (Long) list.stream().filter(item -> item != null)
+            .reduce(0L, (x, y) -> Math.addExact(x.longValue(), y.longValue()));
+
+
+    public static void register(ScalarFunctionModule module) {
+
+        module.register(
+            Signature.scalar(
+                NAME,
+                new ArrayType(DataTypes.NUMERIC).getTypeSignature(),
+                DataTypes.NUMERIC.getTypeSignature()
+            ),
+            ArraySumFunction::new
+        );
+
+        for (var supportedType : DataTypes.NUMERIC_PRIMITIVE_TYPES) {
+            DataType inputDependantOutputType = DataTypes.LONG;
+            if (supportedType == DataTypes.FLOAT || supportedType == DataTypes.DOUBLE) {
+                inputDependantOutputType = supportedType;
+            }
+
+            module.register(
+                Signature.scalar(
+                    NAME,
+                    new ArrayType(supportedType).getTypeSignature(),
+                    inputDependantOutputType.getTypeSignature()
+                ),
+                ArraySumFunction::new
+            );
+        }
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private ArraySumFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        returnType = (DataType<R>) signature.getReturnType().createType();
+        var argumentType = boundSignature.getArgumentDataTypes().get(0);
+        this.ZERO = returnType.implicitCast(0);
+
+        if (returnType == DataTypes.FLOAT) {
+            summationFunction = floatFunction;
+        } else if (returnType == DataTypes.DOUBLE) {
+            summationFunction = doubleFunction;
+        } else if (returnType == DataTypes.NUMERIC) {
+            summationFunction = numericFunction;
+        } else {
+            summationFunction = longFunction;
+        }
+
+        ensureInnerTypeIsNotUndefined(boundSignature.getArgumentDataTypes(), signature.getName().name());
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public R evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input[] args) {
+        List<T> values = (List) args[0].value();
+        if (values == null || values.isEmpty()) {
```

Comment:
For most scalars the behavior is that if the incoming value is `null`, the output is also `null`. Not sure if we should have a implicit `null -> 0` conversion here.

We probably also want to have `array_sum` return the same result as `sum()` on an unnested array would return, so if all values of the array are `null` then the result should probably also be null:

```
cr> select sum(x) from unnest([null, null, null]::int[]) as t (x);
+--------+
| sum(x) |
+--------+
|   NULL |
+--------+
SELECT 1 row in set (0.007 sec)
cr> select array_sum([null, null, null]::int[]);
+-----------+
| 0::bigint |
+-----------+
|         0 |
+-----------+
```

Reply:
Thx, fixed it and updated test.